src/librustc_middle/util/common.rs

```diff
@@ -67,3 +69,28 @@ pub fn indenter() -> Indenter {
     debug!(">>");
     Indenter { _cannot_construct_outside_of_this_module: () }
 }
+
+pub trait TargetExt {
+    // We need to reference `TyCtxt`, so this method cannot go in `rustc_hir`
+    fn from_impl_item<'tcx>(tcx: TyCtxt<'tcx>, impl_item: &ImplItem<'_>) -> Target {
+        match impl_item.kind {
+            ImplItemKind::Const(..) => Target::AssocConst,
+            ImplItemKind::Fn(..) => {
+                let parent_hir_id = tcx.hir().get_parent_item(impl_item.hir_id);
+                let containing_item = tcx.hir().expect_item(parent_hir_id);
+                let containing_impl_is_for_trait = match &containing_item.kind {
+                    ItemKind::Impl { ref of_trait, .. } => of_trait.is_some(),
+                    _ => bug!("parent of an ImplItem must be an Impl"),
+                };
+                if containing_impl_is_for_trait {
+                    Target::Method(MethodKind::Trait { body: true })
+                } else {
+                    Target::Method(MethodKind::Inherent)
+                }
+            }
+            ImplItemKind::TyAlias(..) | ImplItemKind::OpaqueTy(..) => Target::AssocTy,
+        }
+    }
+}
+
+impl TargetExt for Target {}
```

Comment:
This is not the module for this. `rustc_middle/hir` should exist, you can put it anywhere in there.